### PR TITLE
perf(compiler): consume value-range tags in EVM bool/compare/bitwise lowering

### PR DESCRIPTION
**Pull request by abmcar**
Consume existing `ValueRange` tags in four EVM lowering paths so operands already proven to fit in u64 skip full 4-limb sequences. Overall fast-path hit rate rises 78.72% -> 80.02% on the EEST Cancun suite, with no measured performance regression and 21/21 differential fixtures passing.

## What

The multipass JIT range analyzer already proves many operands fit in u64, but several lowering paths ignore that proof and emit full 4-limb sequences. This PR makes four lowerings consume the existing `ValueRange` tags:

1. **ISZERO**: deferred zero-test operands now carry the base value's range and tag their 0/1 result as u64. Materialization folds only the limbs the range allows (u64 base -> single-limb test instead of a 4-limb OR-fold).
2. **JUMPI**: a deferred zero-test condition is fused into the branch — no materialize-then-refold round trip — and non-deferred conditions fold only range-live limbs. The common `LT; ISZERO; JUMPI` loop-exit shape previously paid a full compare, a 4-limb OR-fold + cmp + select materialization, and a second 4-limb OR-fold + cmp + select at the branch.
3. **OR/XOR**: new non-constant narrow paths. Both operands proven u64 -> a single i64 op with zeroed upper limbs (result tagged u64); exactly one side u64 -> low-limb op with the wide side's upper limbs passed through (identity), mirroring the existing const-u64 path.
4. **SLT/SGT vs a u64 constant**: new fast paths (`handleCompareSltRhsU64` / `handleCompareSgtRhsU64`). A u64 constant has zero upper limbs, so it is a non-negative signed-256 value; the compare reduces to sign-bit test + upper-limb OR + an unsigned limb0 compare, with the same three range tiers as the unsigned helpers. Both constant positions are dispatched (`c <s x ⟺ x >s c`).

## Why

Real-mainnet-load profiling showed these ops taking the full path while their operands were already statically proven narrow: 58.5% of ISZERO full-path executions had a proven-u64 operand, 51.8% of OR full-path executions had at least one proven-u64 side, and 22.4% of SLT full-path executions compared against a u64 constant. These are lowering-side gaps — no new analysis precision is required to close them.

## Soundness

- All new paths only narrow the *read width* where a `ValueRange` tag already proves upper limbs are semantically zero; no new tag sources are introduced (change 5's tag is justified by the function's structural zero-fill).
- SLT/SGT: operands with limb0 in `[2^63, 2^64-1]` are positive 256-bit values; the new paths use unsigned predicates on limb0 and decide negative inputs from limb3's sign bit. The truth tables are verified across the boundary set (2^63, 2^64, 2^128, 2^192, 2^255, -1, equal values, c=0, c>=2^63).
- Builder result ranges remain equal to or wider than the analyzer's transfer rules for every touched op (no builder-narrower-than-truth state).
- JUMPI now depends on the range-contract invariant (an over-narrow upstream tag would mis-branch rather than merely run slower); this is documented at the fold site.

## Diff scope

This diff includes the base-range plumbing (`DeferredBaseRange`) and the 1/2/4-limb narrow fold at materialization and JUMPI fusion. Two tags that previously sat here are excluded because they landed upstream independently: the context-size producer tag (now #532) and the U64 tag on the deferred zero-test *result* (now #524).

## Verification

- 21 new differential fixtures (`tests/evm_asm/`, committed) cover each new path's narrow and full sides with boundary values including 2^64, 2^128, 2^192, -1, high-sparse, and the limb0-MSB unsigned-predicate hard gate. A new `EVMRangeNarrowingDifferentialTest` suite asserts interpreter and multipass outputs match byte-for-byte and that multipass actually JIT-compiled: 21/21 pass.
- multipass evmone-unittests: 223/223; multipass evmone-statetest `-k fork_Cancun`: 2723/2723; ctest: 11/11; golden `.easm` suite: 178/178 (no regressions); `tools/format.sh check` clean; no new build warnings.
- Re-run on the rebased branch: differential 21/21, multipass evmone-unittests 223/223, statetest fork_Cancun 2723/2723.

## Measurements

Fast-path hit rate, paired site-weighted measurement on the EEST Cancun suite (28,109 shared compiled sites, measured with an instrumentation-only tap branch not included in this PR). The EEST measurement predates the #522/#524/#530/#532 merges; the ADD row's 17 sites are attributable to the context-size tag landed as #532. Transition column reads as "<count> sites: <full path> -> <narrowed path>".

| op | base | this PR | delta | transitions |
|---|---:|---:|---:|---|
| ISZERO | 23.2% | 85.9% | +62.7pp | 316 sites: full 4-limb -> single-limb u64 |
| SGT | 80.6% | 95.0% | +14.4pp | 20 sites: full 4-limb -> u64-constant compare |
| SLT | 61.5% | 66.7% | +5.2pp | 9 sites: full 4-limb -> u64-constant compare |
| ADD | 75.1% | 75.6% | +0.5pp | 17 sites: full 4-limb -> u128 narrow |
| overall | 78.72% | 80.02% | **+1.29pp** | +364 sites, zero reverse transitions |

Performance (evmone-bench, 27-bench sweep, multipass, vs upstream/main baseline): median delta -0.23%, within run-to-run variance (about ±2%). First-pass outliers re-measured with 15 repetitions all resolved to noise; the largest and most stable benchmark (snailtracer, cv 1.5%) improved consistently across two independent runs (-1.1% / -1.3%). No regressions.

## Known limitation

With `ZEN_ENABLE_EVM_STACK_SSA_LIFT=ON` (default OFF, CI OFF), `getOperandIdentityKey()` in the lifted-stack path does not handle deferred operands; a live-out deferred zero-test can hit an assertion there. This exposure predates this PR (the lifter is untouched and deferred operand lifetimes are unchanged) and is noted in the change document for the stack-lift follow-up.

Change document: `docs/changes/2026-06-10-evm-range-lowering-gaps/README.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/compiler/evm_frontend/evm_mir_compiler.cpp [hunk 1]
```diff
@@ -3190,6 +3227,113 @@ EVMMirBuilder::handleCompareGtRhsU64(const Operand &LHSOp, uint64_t RhsU64) {
...
+  U256Inst LHS = extractU256Operand(LHSOp);
+  MType *MirI64Type =
+      EVMFrontendContext::getMIRTypeFromEVMType(EVMType::UINT64);
+  MInstruction *Zero = createIntConstInstruction(MirI64Type, 0);
+
+  MInstruction *RhsVal = createIntConstInstruction(MirI64Type, RhsU64);
+  auto GtPred = CmpInstruction::Predicate::ICMP_UGT;
+  MInstruction *LowGt = createInstruction<CmpInstruction>(
+      false, GtPred, &Ctx.I64Type, LHS[0], RhsVal);
+
```
[the remark this answers is not in the document]
Comment (abmcar) [reply src/compiler/evm_frontend/evm_mir_compiler.cpp:3423]: The two signed helpers deliberately mirror the structure of the existing unsigned pair (`handleCompareLtRhsU64`/`handleCompareGtRhsU64`) so all four const-compare helpers stay structurally parallel and reviewable side by side — same range tiers, same select assembly. Parametrizing only the signed pair would break that symmetry, and consolidating all four is a refactor beyond this PR's scope. Agreed it becomes mandatory if more signed-compare fast paths are added; noting that as the trigger.

**Comment by abmcar** [comment]
The differential test suite and its fixtures have moved out of this PR into #539, which consolidates the interp-vs-multipass differential coverage from all three optimization PRs into a dedicated test target. This PR is now code + docs only; #539 carries the tests and can merge independently in any order.
